crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -72,24 +95,35 @@ impl<C: BodiesClient> ConcurrentDownloader<C> {
     }
 
     /// Fetch a single block body.
-    async fn fetch_body(
-        &self,
-        block_number: BlockNumber,
-        header_hash: H256,
-    ) -> RequestResult<(BlockNumber, H256, BlockBody)> {
-        let mut response = self.client.get_block_body(vec![header_hash]).await?;
-        Ok((block_number, header_hash, response.1.remove(0))) // TODO:
+    async fn fetch_body(&self, header: &SealedHeader) -> RequestResult<BlockLocked> {
+        let (peer_id, mut response) =
+            self.client.get_block_body(vec![header.hash()]).await?.split();
+
+        let body = response.remove(0);
+        let block = BlockLocked {
+            header: header.clone(),
+            body: body.transactions,
+            ommers: body.ommers.into_iter().map(|header| header.seal()).collect(),
+        };
+        if self.consensus.pre_validate_block(&block).is_err() {
+            self.client.penalize(peer_id);
+        }
+        Ok(block)
```

Comment:
There's a bug here, we should return err block, not ok(block)

Reply:
fixed